### PR TITLE
Add flags to cabana to enable hypre and heffte when they are part of …

### PR DESCRIPTION
**Pull request by patrickb314**
…spec. Also add googletest to build dependencies 

Please review and approve if appropriate, @junghans

### CODE REVIEW

File: var/spack/repos/builtin/packages/cabana/package.py [hunk 1]
```diff
@@ -67,6 +68,13 @@ def cmake_args(self):
             '-DBUILD_SHARED_LIBS=%s' % (
                 'On' if '+shared'  in self.spec else 'Off')
         ]
+
+        # Enable Cabana submodules based on flags above
```
Comment (junghans) [inline var/spack/repos/builtin/packages/cabana/package.py:78]: These don't hurt, but I am wondering if they are actually needed...

Isn't CMake detecting them automatically?

Comment (junghans) [reply var/spack/repos/builtin/packages/cabana/package.py:78]: Actually the options are called `Cabana_REQUIRE_<foo>`, the `ENABLE` options are internal and set to whether the package was found automatically:
https://github.com/ECP-copa/Cabana/blob/master/CMakeLists.txt#L84

Comment (patrickb314) [reply var/spack/repos/builtin/packages/cabana/package.py:78]: It wasn't for me, no. Cabana cmake doesn't look for hypre and then conditionally enable hypre and heffte support. As a result, Cajita_Config.hpp doesn't include the hypre include file or hypre support if this isn't enabled.

Comment (patrickb314) [reply var/spack/repos/builtin/packages/cabana/package.py:78]: Hmmm. Let me try again, then. If I just added +hypre to the spec for my spack environment, concretized it, and then did an install, Cabana still didn't have HYPRE support enabled. I'll see if I can track it down better.

Comment (junghans) [reply var/spack/repos/builtin/packages/cabana/package.py:78]: @streeve do you know?

Comment (streeve) [reply var/spack/repos/builtin/packages/cabana/package.py:78]: I was able to reproduce the issue. Adding the `Cabana_REQUIRE_` flags below will give clear error messages. It looks like there's an issue with the spack required heffte/hypre versions and the actual CMake required versions

Comment (junghans) [reply var/spack/repos/builtin/packages/cabana/package.py:78]: I think we need to depend on `hypre-cmake+shared`.

**Comment by junghans** [comment]
@patrickb314 can you try my fix?

File: var/spack/repos/builtin/packages/cabana/package.py (hunk 1 above)
Comment (streeve) [inline var/spack/repos/builtin/packages/cabana/package.py:78]:
```suggestion
        # Require Cabana submodules based on flags above
        if '+mpi' in self.spec:
            options.append('-DCabana_REQUIRE_MPI=ON')
        if '+arborx' in self.spec:
            options.append('-DCabana_REQUIRE_ARBORX=ON')
```

File: var/spack/repos/builtin/packages/cabana/package.py [hunk 2]
```diff
@@ -54,8 +55,8 @@ class Cabana(CMakePackage):
                 _kk_spec = 'kokkos{0}+{1}'.format(_kk_version, _backend)
             depends_on(_kk_spec, when='@{0}+{1}'.format(_version, _backend))
     depends_on("arborx", when="@0.3.0:+arborx")
-    depends_on("heffte@2.0:", when="@0.4.0:+heffte")
-    depends_on("hypre-cmake@2.22.0:", when="@0.4.0:+hypre")
+    depends_on("heffte+shared@2.0:", when="@0.4.0:+heffte")
```
Comment (streeve) [inline var/spack/repos/builtin/packages/cabana/package.py]:
```suggestion
    # FIXME: Loosen restrictions when fixed in heffte
    depends_on("heffte+shared@2.0", when="@0.4.0+heffte")
    depends_on("heffte+shared@2.1", when="@master+heffte")
```

File: var/spack/repos/builtin/packages/cabana/package.py [hunk 3]
```diff
@@ -54,8 +55,8 @@ class Cabana(CMakePackage):
...
+    depends_on("hypre-cmake+shared@2.22.0:", when="@0.4.0:+hypre")
```
Comment (streeve) [inline var/spack/repos/builtin/packages/cabana/package.py]:
```suggestion
    depends_on("hypre-cmake+shared@2.22.0:", when="@0.4.0+hypre")
    depends_on("hypre-cmake+shared@2.22.1:", when="@master+hypre")
```

**Review by streeve — CHANGES_REQUESTED** [review]
Thanks for fixing!

File: var/spack/repos/builtin/packages/cabana/package.py [hunk 4]
```diff
@@ -54,8 +55,12 @@ class Cabana(CMakePackage):
...
             depends_on(_kk_spec, when='@{0}+{1}'.format(_version, _backend))
     depends_on("arborx", when="@0.3.0:+arborx")
-    depends_on("heffte@2.0:", when="@0.4.0:+heffte")
-    depends_on("hypre-cmake@2.22.0:", when="@0.4.0:+hypre")
+    depends_on("hypre-cmake@2.22:", when="@0.4.0:+hypre")
+    # Heffte pinned at 2.0.0 for now because its cmakefiles can't roll forward
+    # compatibilty from 2.0.0 to later minor versions and cabana cmakefiles are
+    # currently requesting version 2.0.0. Will be fixed for later cabana versions
+    # as cabana requested version number and heffte version compatibility changes
+    depends_on("heffte@2.0.0", when="@0.4.0:+heffte")
```
Comment (junghans) [inline var/spack/repos/builtin/packages/cabana/package.py]:
```suggestion
    depends_on("heffte@2.0.0", when="@0.4.0+heffte")
    depends_on("heffte@2.1.0", when="@master+heffte")
```

**Review by junghans — APPROVED** [review]

**Comment by junghans** [comment]
@patrickb314 ready to merge?

**Comment by patrickb314** [comment]
Yes, I think we're set.

File: var/spack/repos/builtin/packages/cabana/package.py (hunk 1 above)
Comment (streeve) [reply var/spack/repos/builtin/packages/cabana/package.py:78]: This isn't necessary since we don't have as strict version requirements for these, but would provide more consistent behavior
